util/terraform.go

```diff
@@ -3,32 +3,55 @@ package util
 import (
 	"fmt"
 	"io/ioutil"
+	"os"
+	"path"
 	"path/filepath"
+	"regexp"
 	"strings"
 
+	"github.com/fatih/color"
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/hclparse"
 	"github.com/hashicorp/terraform/configs"
 
 	"github.com/coveooss/gotemplate/v3/collections"
 	"github.com/coveooss/gotemplate/v3/json"
 	"github.com/coveooss/gotemplate/v3/template"
+	"github.com/coveooss/gotemplate/v3/utils"
 	"github.com/coveooss/gotemplate/v3/yaml"
+	"github.com/coveooss/multilogger"
 	"github.com/coveooss/multilogger/errors"
 	"github.com/sirupsen/logrus"
 )
 
+type (
+	cachedDefaultVariables struct {
+		values map[string]interface{}
+		all    map[string]*configs.Variable
+		err    error
+	}
+)
+
+var (
+	reTfVariables = regexp.MustCompile(`(?ms)^variable\s+".*?"\s+{(\s*\n.*?^}|\s*}$)`)
+	cacheDefault  = make(map[string]*cachedDefaultVariables)
+)
+
 // LoadDefaultValues returns a map of the variables defined in the tfvars file
-func LoadDefaultValues(folder string) (importedVariables map[string]interface{}, allVariables map[string]*configs.Variable, err error) {
-	var terraformConfig *configs.Module
-	parser := configs.NewParser(nil)
-	if terraformConfig, err = parser.LoadConfigDir(folder); err != nil && err.(hcl.Diagnostics).HasErrors() {
-		err = fmt.Errorf("caught errors while trying to load default variable values from %s:\n%w", folder, errors.Array(err.(hcl.Diagnostics).Errs()))
-		return
+func LoadDefaultValues(folder string, logger *multilogger.Logger) (importedVariables map[string]interface{}, allVariables map[string]*configs.Variable, err error) {
```

Comment:
Why cache? Do we really load default values more than once in a process? When?

Reply:
We were doing twice when I traced the application.

Reply:
In fact, we call it on the main folder (indirectly through `importDefaultVariables`:

```golang
// Retrieve the default variables from the terraform files
if err = importDefaultVariables(terragruntOptions, terragruntOptions.WorkingDir); stopOnError(err) {
	return
}
```

Then, we call it again on each repository (including subfolders)

```golang
// Export Terragrunt variables to the paths defined in export_variables blocks
for _, folder := range foldersWithTerraformFiles {
	var existingVariables map[string]*configs.Variable
	_, existingVariables, err = util.LoadDefaultValues(folder, terragruntOptions.Logger)
	if stopOnError(err) {
		return err
	}
	if err = conf.ExportVariables(existingVariables, folder); stopOnError(err) {
		return
	}
}
```

Reply:
Different folders though. They don't go in the same cache

Reply:
The main folder is processed twice. I can add a parameter telling to use cache or not to avoid keeping useless element in cache. But the memory is not really an issue due to the nature of the program.

Reply:
Ah, no it's fine. Didn't catch that it ran twice on the main folder

Reply:
```text
[terragrunt] 2020/09/23 08:18:41.156  17.2s (10.4s) INFO     Import bootstrap Terraform files to .
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU
[terragrunt] 2020/09/23 08:18:41.262  17.3s (106ms) INFO     Import bootstrap Terraform files to buckets, empty_module, external_peering, key_pair, kms, nat_gateway, nat_instance, r53, subnet, subnet_old, vpc
[terragrunt] 2020/09/23 08:18:48.815  24.8s (7.59s) INFO     File(s) modified by go template: _inputs.tf, _locals.tf, _outputs.tf, certificates.tf, common_policies.tf, dep_role.tf, ds.tf, k8s_base.tf, legacy_vpc_hipaa.tf, main.tf, parameter_store.tf, shared_secrets.tf, sqs_vpc_endpoint.tf, uam_route_table.tf, _bootstrap_providers.tf, subnets.tf, buckets/_interfaces.tf, buckets/backup.tf, buckets/binaries.tf, buckets/buildsystem.tf, buckets/config.tf, buckets/customerdata.tf, buckets/global.tf, buckets/permanent_customerdata.tf, buckets/policy.tf, buckets/public_content.tf, buckets/public_resource.tf, buckets/security.tf, buckets/_bootstrap_providers.tf, empty_module/_bootstrap_providers.tf, external_peering/peering_provider.tf, external_peering/vpc_peering.tf, external_peering/_bootstrap_providers.tf, key_pair/_bootstrap_providers.tf, kms/_bootstrap_providers.tf, nat_gateway/_bootstrap_providers.tf, nat_instance/main.tf, nat_instance/_bootstrap_providers.tf, r53/_outputs.tf, r53/zone.tf, r53/_bootstrap_providers.tf, subnet/_bootstrap_providers.tf, subnet_old/_bootstrap_providers.tf, vpc/main.tf, vpc/peering.tf, vpc/_bootstrap_providers.tf
Cache used /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/buckets
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/empty_module
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/external_peering
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/key_pair
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/kms
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/nat_gateway
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/nat_instance
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/r53
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/subnet
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/subnet_old
Original call /var/tgf/terragrunt/h7cVOGpWALFRo00ZTfcSpc9A4kw/_4A9DFGd5tR8McMRSEXf9YBjtoU/vpc
```

Reply:
Too late, I did it :p